### PR TITLE
Update k8s docs for Vault as a Secrets Backend

### PR DESCRIPTION
**Pull request by jmurret**
This all of the current Vault changes except for the WAN Federation guide.

The main goal of this PR was to introduce adding support for bootstrap token, replication token, snapshot agent config, and enterprise license.

The other goal was to try to break up the various commands in to some further buckets. Integration requires a large run list of commands that is a little hard to grok everything that is occurring and why for those not intimately familiar with it.  I broke this documentation down into two sections:
- Systems Integration - what needs to be turned and configured so that the systems are aware of each other and ready to store and read data
- Data Integration - what needs to be done to store data and configure authorization so Consul on Kubernetes can read a specific secret.

I plan to add Wan Federation documentation as an end to end example at the same level as Systems Integration and Data Integration in the below nav tree.  Also WAN Federation documentation will use the same general steps structure that is used in the Systems Integration and Data Integration Overviews.

Screen Shot 2022-04-04 at 4 24 26 PM

### CODE REVIEW

File: website/content/docs/k8s/installation/vault/index.mdx [hunk 1]
```diff
@@ -11,105 +11,28 @@ By default, Consul Helm chart will expect that any credentials it needs are stor
 As of Consul 1.11 and Consul Helm chart v0.38.0, we integrate more natively with Vault making it easier
 to use Consul Helm chart with Vault as the secrets storage backend.
 
-At a high level, there are two points of integration with Vault:
-- **Gossip encryption** - The encryption key for gossip communication is stored in Vault.
-- **TLS certificates and keys**:
-  - **Consul Server TLS credentials** - TLS certificate and key for the Consul server is stored in Vault and issued from Vault.
-  - **Service Mesh and Consul client TLS credentials** - Consul uses Vault as the provider for mTLS certificates and keys for the service mesh services
-  and TLS certificates and keys for the Consul clients.
+The following secrets can be stored in Vault KV secrets engine, which is meant to handle arbitraty secrets:
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/index.mdx:14]:
```suggestion
The following secrets can be stored in Vault KV secrets engine, which is meant to handle arbitrary secrets:
```

File: website/content/docs/k8s/installation/vault/systems-integration.mdx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  - Installing the Vault Injector within the Consul data center installation
+  - Configuring a Kubernetes Auth Method in Vault to authenticate and authorize operations from the Consul data center
+  - Enable Vault as the secrets backend in the Consul data center
+
+## One time set up on Vault
+### Vault KV Secrets Engine - Version 2
+The following secrets can be stored in Vault KV secrets engine, which is meant to handle arbitrary secrets:
+- ACL Bootstrap token ([`global.acls.bootstrapToken`](/docs/k8s/helm#v-global-acls-bootstraptoken))
+- Gossip encryption key ([`global.gossipEncryption`]())
+- Enterprise license ([`global.enterpriseLicense]())
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/systems-integration.mdx:25]:
```suggestion
- Enterprise license ([`global.enterpriseLicense`]())
```

File: website/content/docs/k8s/installation/vault/systems-integration.mdx [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+### Vault PKI Engine
+
+The Vault PKI Engine must be enabled in order to leverage Vault for issuing Consul Server TLS certificates. More details for configuring the PKI Engine is found in [Bootstrapping the PKI Engine](/docs/k8s/installation/vault/data-integration/server-tls#bootstrapping-the-pki-engine) under the Server TLS section.
+
+```shell-session
+$ vault secrets enable pki
+```
+
+## Set up per Consul data center
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/systems-integration.mdx:43]:
```suggestion
## Set up per Consul Datacenter
```

Datacenter is a proper noun for Consul representing a cluster deployed within a region.

File: website/content/docs/k8s/installation/vault/systems-integration.mdx [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+A minimal valid installation of Vault Kubernetes must include the Agent Injector which is utilized for accessing secrets from Vault. Vault servers could be deployed external to Vault on Kubernetes as described via the `externalvaultaddr` value in the Vault Helm Configuration.
+
+```shell
+$ cat <<EOF >> vault-injector.yaml
+# vault-injector.yaml
+server:
+  enabled: false
+injector:
+  enabled: true
+  externalVaultAddr: $VAULT_ADDR
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/systems-integration.mdx:87]:
```suggestion
  externalVaultAddr: ${VAULT_ADDR}
```

File: website/content/docs/k8s/installation/vault/systems-integration.mdx [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    token_reviewer_jwt="$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" \
+    kubernetes_host="https://$KUBE_API_URL:443" \
+    kubernetes_ca_cert=@/var/run/secrets/kubernetes.io/serviceaccount/ca.crt
+```
+
+#### Enable Vault as the secrets backend in the Consul data center
+Finally, you will configure the Consul on Kubernetes helm chart for the data center to expect to receive the following values (if you have configured them) to be retreived from Vault:
+- ACL Bootstrap token ([`global.acls.bootstrapToken`](/docs/k8s/helm#v-global-acls-bootstraptoken))
+- Gossip encryption key ([`global.gossipEncryption`]())
+- Enterprise license ([`global.enterpriseLicense]())
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/systems-integration.mdx:130]:
```suggestion
- Enterprise license ([`global.enterpriseLicense`]())
```

File: website/content/docs/k8s/installation/vault/data-integration/index.mdx [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  1. Create Vault Kubernetes auth roles that link the policy to each Consul on Kubernetes service account that requires access.
+  1. Configure the Vault Kubernetes auth role in the Consul on Kubernetes helm chart.
+
+For example, at a high level, the steps for [Gossip encryption key](/docs/k8s/installation/vault/data-integration/gossip) would be:
+- One time set up in Vault
+  1. Store the secret in Vault.
+      - You save the gossip encryption key in Vault at the path `secret/consul/gossip`.
+  1. Create a Vault policy that authorizes the desired level of access to the secret.
+      - You create a Vault policy that you name `gossip-policy` which allows `read` access to the path `secret/consul/gossip`.
+- Set up per Consul data center
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/data-integration/index.mdx:28]:
```suggestion
- Set up per Consul Datacenter
```

File: website/content/docs/k8s/installation/vault/data-integration/server-tls.mdx [hunk 7]
```diff
@@ -42,7 +40,27 @@ which also uses an intermediate signing authority.
 
   -> **Note:** Where `common_name` is comprised of combining `global.datacenter` dot `global.domain`.
 
-### Create Vault Policies for the Server TLS Certificates
+## Overview
+To use an Vault as the Server TLS Certificate Provider on Kubernetes, we will need to modify the steps outlined in the [Data Integration](/docs/k8s/installation/vault/data-integration) section:
+
+- One time set up in Vault
+  1. ~~Store the secret in Vault.~~
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/data-integration/server-tls.mdx:47]:
```suggestion
```


Perhaps we can remove this from the list but still mention it below in the case of posterity to previous setup instructions.

Comment (david-yu) [reply website/content/docs/k8s/installation/vault/data-integration/server-tls.mdx:47]: I found it a looking a little odd that this was striked out.

File: website/content/docs/k8s/installation/vault/data-integration/connect-ca.mdx [hunk 8]
```diff
@@ -11,14 +11,36 @@ description: >-
...
+Prior to setting up the data integration between Vault and Consul on Kubernetes, you will need to have:  
+1. Read and completed the steps in the [Systems Integration](/docs/k8s/installation/vault/systems-integration) section of [Vault as a Secrets Backend](/docs/k8s/installation/vault).
+2. Read the [Data Integration Overview](/docs/k8s/installation/vault/systems-integration) section of [Vault as a Secrets Backend](/docs/k8s/installation/vault).
+
+
+## Overview
+To use an Vault as the Service Mesh Certificate Provider on Kubernetes, we will need to modify the steps outlined in the [Data Integration](/docs/k8s/installation/vault/data-integration) section:
+
+- One time set up in Vault
+  1. ~~Store the secret in Vault.~~
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/data-integration/connect-ca.mdx:25]:
```suggestion
```

Same comment as https://github.com/hashicorp/consul/pull/12691/files#r842222392

File: website/content/docs/k8s/installation/vault/data-integration/gossip.mdx [hunk 9]
```diff
@@ -1,26 +1,35 @@
...
-1. Create policies that will allow Consul client and server to access that key.
-1. Create a Kubernetes auth roles that link policies from step 2 to Kubernetes service accounts of the Consul servers and clients.
+## Overview
+To use a gossip encryption key stored in Vault, we will follow the steps outlined in the [Data Integration](/docs/k8s/installation/vault/data-integration) section:
 
-## Configuring Vault
+- One time set up in Vault
+  1. Store the secret in Vault.
+  1. Create a Vault policy that authorizes the desired level of access to the secret.
+- Set up per Consul data center
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/data-integration/gossip.mdx:21]:
```suggestion
- Set up per Consul datacenter
```

File: website/content/docs/k8s/installation/vault/data-integration/snapshot-agent-config.mdx [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+1. Read and completed the steps in the [Systems Integration](/docs/k8s/installation/vault/systems-integration) section of [Vault as a Secrets Backend](/docs/k8s/installation/vault).
+2. Read the [Data Integration Overview](/docs/k8s/installation/vault/systems-integration) section of [Vault as a Secrets Backend](/docs/k8s/installation/vault).
+
+## Overview
+To use an ACL replication token stored in Vault, we will follow the steps outlined in the [Data Integration](/docs/k8s/installation/vault/data-integration) section:
+
+- One time set up in Vault
+  1. Store the secret in Vault.
+  1. Create a Vault policy that authorizes the desired level of access to the secret.
+- Set up per Consul data center
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/data-integration/snapshot-agent-config.mdx:21]:
```suggestion
- Set up per Consul datacenter
```

File: website/content/docs/k8s/installation/vault/data-integration/bootstrap-token.mdx [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+1. Read and completed the steps in the [Systems Integration](/docs/k8s/installation/vault/systems-integration) section of [Vault as a Secrets Backend](/docs/k8s/installation/vault).
+2. Read the [Data Integration Overview](/docs/k8s/installation/vault/systems-integration) section of [Vault as a Secrets Backend](/docs/k8s/installation/vault).
+
+## Overview
+To use an ACL bootstrap token stored in Vault, we will follow the steps outlined in the [Data Integration](/docs/k8s/installation/vault/data-integration) section:
+
+- One time set up in Vault
+  1. Store the secret in Vault.
+  1. Create a Vault policy that authorizes the desired level of access to the secret.
+- Set up per Consul data center
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/data-integration/bootstrap-token.mdx:21]:
```suggestion
- Set up per Consul datacenter
```

File: website/content/docs/k8s/installation/vault/data-integration/connect-ca.mdx [hunk 12]
```diff
@@ -11,14 +11,36 @@ description: >-
...
+  1. Create a Vault policy that authorizes the desired level of access to the secret.
+- Set up per Consul data center
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/data-integration/connect-ca.mdx:27]:
```suggestion
- Set up per Consul datacenter
```

File: website/content/docs/k8s/installation/vault/data-integration/enterprise-license.mdx [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+1. Read and completed the steps in the [Systems Integration](/docs/k8s/installation/vault/systems-integration) section of [Vault as a Secrets Backend](/docs/k8s/installation/vault).
+2. Read the [Data Integration Overview](/docs/k8s/installation/vault/systems-integration) section of [Vault as a Secrets Backend](/docs/k8s/installation/vault).
+
+## Overview
+To use an enterprise license stored in Vault, we will follow the steps outlined in the [Data Integration](/docs/k8s/installation/vault/data-integration) section:
+
+- One time set up in Vault
+  1. Store the secret in Vault.
+  1. Create a Vault policy that authorizes the desired level of access to the secret.
+- Set up per Consul data center
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/data-integration/enterprise-license.mdx:21]:
```suggestion
- Set up per Consul datacenter
```

File: website/content/docs/k8s/installation/vault/data-integration/replication-token.mdx [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+To find out the service account name of the Consul server,
+you can run the following `helm template` command with your Consul on Kubernetes values file:
+
+- Generate Consul server service account name
+  ```shell-session
+  $ helm template --release-name ${RELEASE_NAME} -s templates/server-acl-init-serviceaccount.yaml hashicorp/consul
+  ```
+
+## Configure the Vault Kubernetes auth role in the Consul on Kubernetes helm chart
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/data-integration/replication-token.mdx:76]:
```suggestion
### Configure the Vault Kubernetes auth role in the Consul on Kubernetes helm chart
```

File: website/content/docs/k8s/installation/vault/data-integration/replication-token.mdx [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+</CodeBlockConfig>
+
+```shell-session
+$ vault policy write replication-token-policy replication-token-policy.hcl
+```
+
+-> **Note:** The secret path referenced by the above Policy will be your `global.acls.replicationToken.secretName` Helm value.
+
+## Set up per Consul data center
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/data-integration/replication-token.mdx:55]:
```suggestion
## Set up per Consul datacenter
```

File: website/content/docs/k8s/installation/vault/data-integration/server-tls.mdx (hunk 7 above)
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/data-integration/server-tls.mdx:47]:
```suggestion
```

File: website/content/docs/k8s/installation/vault/systems-integration.mdx [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+page_title: Vault as Secrets Backend Systems Integration Overview
+description: >-
+  Overview of the systems integration aspects to using Vault as secrets backend for Consul on Kubernetes.
+---
+
+# Vault as the secrets backend - Systems Integration
+
+## Overview
+At a high level, configuring a systems integration of Vault with Consul on Kubernetes consists of the following steps:
+- One time set up on Vault
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/systems-integration.mdx:12]:
```suggestion
## One time set up on Vault
```

File: website/content/docs/k8s/installation/vault/systems-integration.mdx [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+  - Enabling Vault KV Secrets Engine - Version 2 to store arbitrary secrets
+  - Enabling Vault PKI Engine if you are choosing to store and manage either [Consul Server TLS credentials](/docs/k8s/installation/vault/data-integration/server-tls) or [Service Mesh and Consul client TLS credentials](/docs/k8s/installation/vault/data-integration/connect-ca)
+- Set up per Consul datacenter
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/systems-integration.mdx:15]:
```suggestion
## Set up per Consul datacenter
```

File: website/content/docs/k8s/installation/vault/systems-integration.mdx [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+
+Next, you will configure the Vault Kubernetes Auth Method for the datacenter.  You will need to provide it with:
+- token_reviewer_jwt - this a JWT token from the Consul datacenter cluster that the Vault Kubernetes Auth Method will use to query the Consul datacenter Kubernetes API when services in the Consul datacenter request data from Vault.
+- kubernetes_host - this is the URL of the Consul datacenter's Kubernetes API that Vault will query to authenticate the service account of an incoming request from a Consul data ceneter kubernetes service.
+- kubernetes_ca_cert - this is the CA certifcation that is currently being used by the Consul datacenter Kubernetes cluster.
+
+```shell-session
+$ vault write auth/kubernetes/config \
+    token_reviewer_jwt="$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" \
+    kubernetes_host="https://$KUBE_API_URL:443" \
```
Comment (david-yu) [inline website/content/docs/k8s/installation/vault/systems-integration.mdx:122]:
```suggestion
    kubernetes_host="${https://$KUBE_API_URL}:443" \
```

**Review by david-yu — APPROVED** [review]
This is good to go from my side. I'll need to revisit again in the future to perhaps refactor the `helm template` commands to set env variables for use in the `vault write policy` commands. But it looks good so far! Thanks for your work on this, this format looks much more digestible for someone starting off with this feature.

**Review by ishustava — APPROVED** [review]
Love the reorganization of the Vault concepts!!

I'm leaving some edits and questions but won't block the merge. If I find anything else, I'll submit a separate PR to fix because I probably won't have time to review in more detail.

File: website/content/docs/k8s/installation/vault/systems-integration.mdx [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+#### VAULT_ADDR
+
+  - **Recommended value:** Connecting to Port 8200 of 
+    ```shell-session
+    $ export VAULT_ADDR=http://${VAULT_SERVER_HOST}:8200
+    ```
+
+### Install Vault Injector in your Consul k8s cluster
+
+A minimal valid installation of Vault Kubernetes must include the Agent Injector which is utilized for accessing secrets from Vault. Vault servers could be deployed external to Vault on Kubernetes as described via the `externalvaultaddr` value in the Vault Helm Configuration.
```
[the remark this answers is not in the document]
Comment (jmurret) [reply website/content/docs/k8s/installation/vault/systems-integration.mdx:98]: No. This is in the existing docs here: https://www.consul.io/docs/k8s/installation/vault#vault-helm-config

Should we keep this?

[the remark this answers is not in the document]
Comment (ishustava) [reply website/content/docs/k8s/installation/vault/systems-integration.mdx:98]: I think it's ok to keep it, although maybe we can fix the helm value:
```suggestion
A minimal valid installation of Vault Kubernetes must include the Agent Injector which is utilized for accessing secrets from Vault. Vault servers could be deployed external to Vault on Kubernetes with the [`injector.externalVaultAddr`](https://www.vaultproject.io/docs/platform/k8s/helm/configuration#externalvaultaddr) value in the Vault Helm Configuration.
```

File: website/content/docs/k8s/installation/vault/systems-integration.mdx [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+
+Please read [Run Vault on Kubernetes](https://www.vaultproject.io/docs/platform/k8s/helm/run) if instructions on setting up a Vault cluster are needed. 
+
+### Vault KV Secrets Engine - Version 2
+
+The following secrets can be stored in Vault KV secrets engine, which is meant to handle arbitrary secrets:
+- ACL Bootstrap token ([`global.acls.bootstrapToken`](/docs/k8s/helm#v-global-acls-bootstraptoken))
+- Gossip encryption key ([`global.gossipEncryption`]())
+- Enterprise license ([`global.enterpriseLicense`]())
+- ACL Replication token ([`global.acls.replicationToken`](/docs/k8s/helm#v-global-acls-replicationtoken))
```
[the remark this answers is not in the document]
Comment (jmurret) [reply website/content/docs/k8s/installation/vault/systems-integration.mdx:35]: I just created this page and set up all of the appropriate links in the various pages and nav.

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/630282.

**Comment by hc-github-team-consul-core** [comment]
:cherries::white_check_mark: Cherry pick of commit fd08b6aaf62b0a915a45c6c4828fd1d532b6d26b onto `stable-website` succeeded!

**Comment by hc-github-team-consul-core** [comment]
:cherries::white_check_mark: Cherry pick of commit fd08b6aaf62b0a915a45c6c4828fd1d532b6d26b onto `release/1.12.x` succeeded!

**Pull request merged**